### PR TITLE
updating older versioning api labels to specify patching

### PR DESCRIPTION
**Pull request by flossypurse**
Updates the "Workflow Versioning" section of the Workflow Definition concept to explicitly call out Patching and Build Ids as versioning options and links to the corresponding "how-to" sections for each language.

**Preview:** 

- https://temporal-documentation-git-edu-1197.preview.thundergun.io/workflows#workflow-versioning

The "Versioning" guides for each language have also been updated slightly to explicitly have "Patching" headers within the guides.

**Preview:**

- https://temporal-documentation-git-edu-1197.preview.thundergun.io/dev-guide/go/versioning
- https://temporal-documentation-git-edu-1197.preview.thundergun.io/dev-guide/java/versioning
- https://temporal-documentation-git-edu-1197.preview.thundergun.io/dev-guide/python/versioning
- https://temporal-documentation-git-edu-1197.preview.thundergun.io/dev-guide/typescript/versioning

### CODE REVIEW

**Comment by rachfop** [comment]
Python was missed.
https://temporal-documentation-git-edu-1197.preview.thundergun.io/dev-guide/python/versioning

File: docs-src/java/versioning-intro.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+sidebar_label: Patching
+description: Use the Java SDK's versioning features to update Workflow code.
+tags:
+  - java
+  - workflow code
+  - versioning
+  - patching
+---
+
+The Temporal Platform requires that Workflow code is [deterministic](/concepts/what-is-a-workflow-definition#deterministic-constraints) in nature.
```
Comment (djmagee) [inline docs-src/java/versioning-intro.md:13]:
```suggestion
The Temporal Platform requires that Workflow code is [deterministic](/concepts/what-is-a-workflow-definition#deterministic-constraints).
```

File: docs-src/typescript/patching.md [hunk 2]
```diff
@@ -1,42 +1,18 @@
 ---
 id: patching
-title: Patching TypeScript Workflows
-sidebar_label: Patching (Migrating)
-description: Any Workflow code change that affects the order in which commands are generated breaks this assumption. So we have to keep both the old and new code when migrating Workflows while they are still running.
+title: How to patch Workflow code in TypeScript
+description: The TypeScript SDK Patching API let you change Workflow Definitions without causing non-deterministic behavior in current long-running Workflows.
```
Comment (djmagee) [inline docs-src/typescript/patching.md:4]:
```suggestion
description: The TypeScript SDK Patching API lets you change Workflow Definitions without causing non-deterministic behavior in current long-running Workflows.
```

File: docs-src/typescript/patching.md [hunk 3]
```diff
@@ -1,42 +1,18 @@
...
- - patching
- - versioning
+  - typescript
+  - workflow code
+  - patching
+  - versioning
 ---
 
-## Alternatives
+Patching APIs let you change Workflow Definitions without causing [non-deterministic behavior](/concepts/what-is-a-workflow-definition#non-deterministic-change) in current long-running Workflows.
```
Comment (djmagee) [inline docs-src/typescript/patching.md:13]:
```suggestion
The TypeScript SDK Patching API lets you change Workflow Definitions without causing [non-deterministic behavior](/concepts/what-is-a-workflow-definition#non-deterministic-change) in current long-running Workflows.
```

File: docs-src/typescript/patching.md [hunk 4]
```diff
@@ -103,7 +79,7 @@ We explain more in this optional 30 minute introduction: [https://www.youtube.co
 
 </details>
 
-## TypeScript SDK Patching API
+### TypeScript SDK Patching API
 
 In principle, the TypeScript SDK's patching mechanism works in a similar "feature-flag" fashion to the other SDK's, however, the "versioning" API has been updated to a notion of "patching in" code.
```
Comment (djmagee) [inline docs-src/typescript/patching.md:84]:
```suggestion
In principle, the TypeScript SDK's patching mechanism works in a similar "feature-flag" fashion to the other SDKs; however, the "versioning" API has been updated to a notion of "patching in" code.
```

File: docs-src/typescript/patching.md [hunk 5]
```diff
@@ -187,7 +163,7 @@ export async function myWorkflow(): Promise<void> {
 
 <!--SNIPEND-->
 
-### Step 2: Deprecate patch
+#### Step 2: Deprecate patch
 
 Once we know that all Workflows started with `v1` code have completed we can [deprecate the patch](https://typescript.temporal.io/api/namespaces/workflow#deprecatepatch).
```
Comment (djmagee) [inline docs-src/typescript/patching.md:168]:
```suggestion
When we know that all Workflows started with `v1` code have completed, we can [deprecate the patch](https://typescript.temporal.io/api/namespaces/workflow#deprecatepatch).
```

File: docs-src/typescript/patching.md [hunk 6]
```diff
@@ -187,7 +163,7 @@ export async function myWorkflow(): Promise<void> {
...
 Deprecated patches bridge between `v2` and `vFinal` (the end result), they work similarly to regular patches by recording a marker in the Workflow history, this marker does not fail replay when Workflow code does not emit it.
```
Comment (djmagee) [inline docs-src/typescript/patching.md:169]:
```suggestion
Deprecated patches bridge between `v2` and `vFinal` (the end result).
They work similarly to regular patches by recording a marker in the Workflow history.
This marker does not fail replay when Workflow code does not emit it.
```

File: docs-src/typescript/versioning-intro.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+title: Introduction to versioning Workflow code in TypeScript
+sidebar_label: Versioning
+description: Use the TypeScript SDK's versioning features to update Workflow code.
+tags:
+  - typescript
+  - workflow code
+  - versioning
+---
+
+The Temporal Platform requires that Workflow code is [deterministic](/concepts/what-is-a-workflow-definition#deterministic-constraints) in nature.
```
Comment (djmagee) [inline docs-src/typescript/versioning-intro.md:12]:
```suggestion
The Temporal Platform requires that Workflow code is [deterministic](/concepts/what-is-a-workflow-definition#deterministic-constraints).
```

File: docs-src/typescript/versioning-intro.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+Because of that requirement, the Temporal TypeScript SDK offers two dedicated versioning features.
+
+- [Workflow Patching APIs](/java/patching)
+- [Worker Build Ids](/java/how-to-use-worker-versioning-in-java)
+
+### Alternatives
+
+Before you explore dedicated versioning features, check if your needs can be addressed in other ways:
```
Comment (djmagee) [inline docs-src/typescript/versioning-intro.md:20]:
```suggestion
Before you explore dedicated versioning features, check whether your needs can be addressed in other ways:
```

File: docs-src/typescript/versioning-intro.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+- [Version Task Queue](#version-task-queue)
+- [Version Workflow Name](#version-workflow-name)
+
+Both options mean that Workflows running `v1` code will never migrate to `v2` code: they will run `v1` code to completion.
```
Comment (djmagee) [inline docs-src/typescript/versioning-intro.md:25]:
```suggestion
Both options mean that Workflows running `v1` code will never migrate to `v2` code; they will run `v1` code to completion.
```

File: docs-src/typescript/versioning-intro.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+If you would like to update Workflows running `v1` _while they are still running_, you [may need to "patch in" code](#do-i-need-to-patch).
```
Comment (djmagee) [inline docs-src/typescript/versioning-intro.md:26]:
```suggestion
If you would like to update Workflows running `v1` _while they are still running,_ you [might need to "patch in" code](#do-i-need-to-patch).
```

File: docs-src/typescript/versioning-intro.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+#### Version Task Queue
+
+If we're currently running our v1 Workflow code on Workers that poll on `queue1`, we can run v2 Workflow code on Workers that poll on `queue2`:
```
Comment (djmagee) [inline docs-src/typescript/versioning-intro.md:30]:
```suggestion
If you're currently running v1 Workflow code on Workers that poll on `queue1`, you can run v2 Workflow code on Workers that poll on `queue2`:
```

File: docs-src/typescript/versioning-intro.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+1. Leave some Workers running your v1 `Workflow`, on the `queue1` Task Queue.
+1. Change your `Workflow` code and spin up new Workers that are polling a `queue2` Task Queue.
+1. Cut over your Clients to only call `Workflow` on `queue2` from now on.
+1. Remove your v1 Workers when all the v1 Workflows have completed.
+
+#### Version Workflow Name
+
+While versioning the Task Queue is usually easier, we can also create a new version of a Workflow by copying it and changing its name:
```
Comment (djmagee) [inline docs-src/typescript/versioning-intro.md:39]:
```suggestion
Although versioning the Task Queue is usually easier, we can also create a new version of a Workflow by copying it and changing its name:
```

File: docs-src/java/patching.md [hunk 13]
```diff
@@ -1,28 +1,17 @@
...
-
-Because we design for potentially long-running Workflows at scale, versioning with Temporal works differently than with other workflow systems.
-We explain more in this optional 30-minute introduction: [https://www.youtube.com/watch?v=kkP899WxgzY](https://www.youtube.com/watch?v=kkP899WxgzY).
-
-## Java Versioning API
-
-Use the `Workflow.getVersion` function to return a version of the code that should be executed and then use the returned
-value to pick a correct branch. Let's look at an example.
+Use the `Workflow.getVersion` function to return a version of the code that should be executed and then use the returned value to pick a correct branch.
+Let's look at an example.
```
[the remark this answers is not in the document]
Comment (flossypurse) [reply docs-src/java/patching.md:14]: If you have a suggestion, go for it.
This is from the original. Was aiming for minimal rewrites.

**Comment by flossypurse** [comment]
> Python was missed. https://temporal-documentation-git-edu-1197.preview.thundergun.io/dev-guide/python/versioning

@rachfop I intentionally didn't do anything there because that page already explicitly calls out Patching versioning and Worker versioning -

The other 3 didn't have that -

**Review by rachfop — APPROVED** [review]

**Pull request merged**